### PR TITLE
Changes drop function to work in log n time

### PR DESCRIPTION
**Pull request by EleanorWhite**
I think this new version of the drop function is in log n time.

### CODE REVIEW

**Comment by cartazio** [comment]
Cool!
Could you walk me through the logic , either diagrammatically , or through some examples, or sort of inductive proof style reasoning? (I'll have a look see myself tomorrow too )

Also do we have tests for drop to validate correctness? 
I think I see why this is correct, but let's talk it out a little bit!  (Good job though)

@ekmett : this is the pr Nell has put together for that feature you wanted :), what's your take?

**Comment by ekmett** [comment]
There is code in my https://github.com/ekmett/lca package that does this
and an article I wrote up on the topic:

https://www.schoolofhaskell.com/user/edwardk/online-lca

**Comment by cartazio** [comment]
https://github.com/ekmett/lca/blob/master/src/Data/LCA/Online/Monoidal.hs#L242

I see!
Looks nice.  I'll compare the two later today.

**Comment by EleanorWhite** [comment]
Sorry it's taken me so long to respond to this.

The program is the same except for how it splits a binary tree.
For binary trees, if it wants to drop less than half or exactly half the size of the tree, it adds the left subtree to the existing RAList and then tries to split the right subtree. The left tree will be smaller than all the trees currently in the RAList because it would have already dropped any trees smaller than the one currently being split. 
If the program needs to drop more than half the size of the tree, it will ignore the left subtree, and try to split the right subtree.

Is this what you meant, or do you want a different/more extensive explanation?

**Comment by cartazio** [comment]
This is a helpful start!
I've had a busy week (as I'm sure you do too), so I've not yet had the time
to work through this. Holler at me Tomorrrow end of day / this wekeend and
I'll try to draw some pictures for myself to understand the cases.

This write up / explanation is a good start.  It might also be handy to
compare (mayve in a different ticket) what differences our ralist code has
vs the linked monoidal structure in eds LCA code base I linked in the pr or
ticket.  That maybe should be in another ticket or something we email back
and forth and then put the summary somewhere.

But yeah, this summary is helpful for me to load in the right thinking
mode.  Either way, we should make sure we are both confident in the
correctness and complexity bounds before we merge it in,

and I'm sure some cleanup ideas will follow from looking at some other data
structure libraries.

One later thing we should also do is some benchmarking! I'll help us get
setup once we get that far. (But that's off topic for this ticket )

Point being: thanks, this summary will help me think through this better!

**Comment by ekmett** [comment]
Eleanor's understanding matches mine.

We have an invariant, which is that a skew binary random access list is a linked list of trees, where each one is larger than the one before it, except at the head of the list where they can be the same size, right?

So, O(log n) drop proceeds by dropping whole trees from the front of the list, this phase can never violate the invariant, as it never produces any new trees. Afterwards, it starts to descend a tree. Any trees that result from this process will be smaller than the source tree it is carving up, so they are suited to the front of the list ahead of other trees that must be larger than the output of the carving process.

Then the cuts happen either by cutting the top off the tree and keeping both children (which matches the 'you can have two of the smallest tree size' condition, as any subsequent tree after the 'critical tree' must be larger than the ones obtained by cutting up the critical tree as noted above.) or by determining if it has to recurse into the left or right child and repeating the process. If the left child, the result entries will be smaller than the right child, which is smaller than the stuff further to the right in the list. If the right child, the same reasoning holds. The left is dropped, so producing two nodes is perfectly acceptable.

That's it. You write up those cases and turn it into code and it does the only thing it can do that preserves the invariants, doesn't commute or duplicate entries and produces a structure of the right size.

**Comment by cartazio** [comment]
Ok, that nicely crystalizes it for me. Thanks!  I'll merge stuff in when
I'm a a train this Saturday :)

**Comment by cartazio** [comment]
Slowly catching up on stuff. This will be merged this week. Mebe tomorrow!

**Comment by cartazio** [comment]
Derp, this still needs to be merged :)

File: Data/RAList.hs
Comment (cartazio) [inline Data/RAList.hs:235]: hrmm, this is something we should look at, i dont remember writing CPP that looks like that :(

File: Data/RAList.hs
Comment (cartazio) [inline Data/RAList.hs]: these NF data instances look correct, but I think  for new enough GHC / base / deepseq  ghc generics give use these for free?

**Comment by EleanorWhite** [comment]
This is with whnf

**Comment by EleanorWhite** [comment]
[report.html.zip.tar.gz](https://github.com/cartazio/ralist/files/688619/report.html.zip.tar.gz)

**Comment by EleanorWhite** [comment]
[report.txt](https://github.com/cartazio/ralist/files/688634/report.txt)

**Comment by cartazio** [comment]
what these are telling me is that we want to look at the list length and tree depth stats for each of these input sizes

**Comment by EleanorWhite** [comment]
[report.html.tar.gz](https://github.com/cartazio/ralist/files/688665/report.html.tar.gz)

**Comment by EleanorWhite** [comment]
[report.html.tar.gz](https://github.com/cartazio/ralist/files/688669/report.html.tar.gz)

**Comment by cartazio** [comment]
@EleanorWhite i think at some level, we're also getting confused about understanding best vs average vs worst case behavior, but either way, the numbers look sane now :) 

if you want, i'll do a merge of a cleaned up diff / rebased change set myself later today

**Comment by EleanorWhite** [comment]
[reporto0.html.tar.gz](https://github.com/cartazio/ralist/files/690582/reporto0.html.tar.gz)
[report02.html.tar.gz](https://github.com/cartazio/ralist/files/690583/report02.html.tar.gz)
These are the two benchmark results using optimization levels O0 and O2

**Comment by cartazio** [comment]
cool! btw, I generally regard any times under 20-30ns as being so small as to be essentially measurement error / noise. 

the timing differences for the simpledrop codes in the O0 vs O2 cases are interesting, but either way looks like everything has the right order of magnitude of perf for merging in :) 

@EleanorWhite i may do a rebase / compaction of these commits later today if you're ok with that

**Comment by EleanorWhite** [comment]
This is the benchmark results using both Safe Haskell and O2 optimization.
[reportO2Safe.html.tar.gz](https://github.com/cartazio/ralist/files/690626/reportO2Safe.html.tar.gz)

**Comment by cartazio** [comment]
@EleanorWhite ok thats interesting, it kinda validates my hypothesis that some list optimizations are firing to make simpleDrop code faster in some cases :)

**Comment by EleanorWhite** [comment]
This is the benchmark results when the library has optimization level O2, and the benchmarking file has O0.
[reportO2O0.html.tar.gz](https://github.com/cartazio/ralist/files/690646/reportO2O0.html.tar.gz)

**Comment by cartazio** [comment]
great work! now lets just do that basic cleanup and we can circle back to analyzing the perf and doing a detailed profiling etc

**Comment by cartazio** [comment]
(after this pr gets merged in )

File: Data/RAList.hs
Comment (cartazio) [inline Data/RAList.hs:1]: We don't need safe in the final pr, but I'll remove this after the merge :)

File: ralist.cabal
Comment (cartazio) [inline ralist.cabal:49]: Probably good to have both be O2, though I should audit the core for differences etc

File: Data/RAList.hs
Comment (cartazio) [reply Data/RAList.hs:1]: In this case safe might be ok. But for library code often simpler to mark as trustworthy

**Review by cartazio — APPROVED** [review]

**Comment by cartazio** [comment]
cherry picked onto master as 
720a57797da52a11b25a19574ea709ecd966ad71
and 
4a1bb66f969a19f06ab4469205720d44a3a7eec3

 merged in, thanks! closing for now :)

**Pull request closed without merging**
